workers/unity/Packages/io.improbable.gdk.core/Systems/EntitySystem.cs

```diff
@@ -1,61 +1,127 @@
+            public JobHandle JobHandle { get; set; }
 
-        public List<EntityId> GetEntitiesRemoved()
-        {
-            return entitiesRemoved;
-        }
-
-        internal void ApplyDiff(ViewDiff diff)
-        {
-            using (applyDiffMarker.Auto())
+            public EntityCollection(int size)
             {
-                entitiesAdded.Clear();
-                entitiesRemoved.Clear();
+                EntityIds = new NativeArray<EntityId>(size, Allocator.Persistent);
+                EntityCount = 0;
+                JobHandle = default;
+            }
 
-                // todo decide on a container and remove this
-                foreach (var entityId in diff.GetEntitiesAdded())
+            public NativeSlice<EntityId> Slice
+            {
+                get
                 {
-                    entitiesAdded.Add(entityId);
+                    JobHandle.Complete();
+                    return new NativeSlice<EntityId>(EntityIds, 0, EntityCount);
                 }
+            }
 
-                foreach (var entityId in diff.GetEntitiesRemoved())
+            public void EnsureSize(ref EntityQuery query)
+            {
+                EntityCount = query.CalculateEntityCount();
+                if (EntityIds.Length >= EntityCount)
                 {
-                    entitiesRemoved.Add(entityId);
+                    return;
                 }
 
-                if (entitiesAdded.Count != 0 || entitiesRemoved.Count != 0)
+                EntityIds.Dispose();
+                EntityIds = new NativeArray<EntityId>(EntityCount << 1, Allocator.Persistent);
+            }
+
+            public void Dispose()
+            {
+                if (EntityIds.IsCreated)
                 {
-                    ViewVersion += 1;
+                    EntityIds.Dispose();
                 }
             }
         }
 
+        private EndSimulationEntityCommandBufferSystem bufferSystem;
+        private EntityCollection added;
+        private EntityQuery addedQuery; // This query needs to be a member of SystemBase
+        private EntityCollection removed;
+        private EntityQuery removedQuery; // This query needs to be a member of SystemBase
+        public int ViewVersion { get; private set; }
+        public NativeSlice<EntityId> EntitiesRemoved => removed.Slice;
+        public NativeSlice<EntityId> EntitiesAdded => added.Slice;
+
+        protected override void OnCreate()
+        {
+            base.OnCreate();
+            bufferSystem = World.GetOrCreateSystem<EndSimulationEntityCommandBufferSystem>();
+            removed = new EntityCollection(1);
+            added = new EntityCollection(1);
+        }
+
         protected override void OnUpdate()
         {
-            entitiesAdded.Clear();
-            entitiesRemoved.Clear();
+            UpdateEntitiesAdded();
+            UpdateEntitiesRemoved();
+
+            Dependency = JobHandle.CombineDependencies(added.JobHandle, removed.JobHandle);
```

Comment:
This code is a little confusing, as it is not clear where the jobhandles come from.
why not have `UpdateEntitiesAdded` and `UpdateEntitiesRemoved` return their job handles, instead of setting them on `added` and `removed`, then do that step here.